File: src/multivariable_calculus_double_integrals_hard.py
"""
multivariable_calculus - double_integrals (hard)
Generated: 2026-02-22T05:39:52.897468
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)

    if problem_type == 1:
        # Reverse order: 0 <= x <= a, x^2 <= y <= cx (compute in original order)
        a = randint(1, 3)
        c = randint(2, 4)
        coeff = nonzero(-2, 2)
        integrand = choice([x*y, x**2 + y, coeff*x*y**2, coeff*x**2*y])
        inner_y = integrate(integrand, (y, x**2, c*x))
        ans = integrate(inner_y, (x, 0, a))
        return problem(
            question=f"Evaluate $\\int_{{0}}^{{{a}}} \\int_{{x^2}}^{{{c}x}} {latex(integrand)} \\, dy \\, dx$",
            answer=ans,
            difficulty=(1650, 1750),
            topic="multivariable_calculus/double_integrals",
            solution=steps(
                f"Inner ($y$ from $x^2$ to ${c}x$): $\\int_{{x^2}}^{{{c}x}} {latex(integrand)} \\, dy = {latex(inner_y)}$",
                f"Outer: $\\int_0^{{{a}}} {latex(inner_y)} \\, dx = {latex(ans)}$"
            ),
            calculator="scientific"
        )

    elif problem_type == 2:
        # Polar coordinates over disk
        r_max = randint(2, 5)
        power = randint(1, 3)
        integrand_cart = (x**2 + y**2)**power
        # polar: r^(2p), with Jacobian r
        r_power = 2*power + 1
        r_integral = Rational(r_max**r_power, r_power) * 2 * pi
        ans = r_integral
        return problem(
            question=f"Evaluate $\\iint_D (x^2+y^2)^{{{power}}} \\, dA$ over the disk $x^2+y^2 \\le {r_max**2}$.",
            answer=ans,
            difficulty=(1600, 1700),
            topic="multivariable_calculus/double_integrals",
            solution=steps(
                f"Polar: $x=r\\cos\\theta$, $y=r\\sin\\theta$, $dA = r \\, dr \\, d\\theta$",
                f"$(x^2+y^2)^{{{power}}} = r^{{{2*power}}}$",
                f"$\\int_0^{{2\\pi}} \\int_0^{{{r_max}}} r^{{{2*power}}} \\cdot r \\, dr \\, d\\theta = \\int_0^{{2\\pi}} \\frac{{{r_max}^{{{2*power+1}}}}}{{{2*power+1}}} d\\theta$",
                f"$= \\frac{{2\\pi \\cdot {r_max**r_power}}}{{{r_power}}} = {latex(ans)}$"
            ),
            calculator="scientific"
        )

    elif problem_type == 3:
        # Region between y = x and y = x^2
        a = randint(1, 2)
        coeff = nonzero(-3, 3)
        func_choice = choice(['x_plus_y', 'xy', 'coeff_x'])
        if func_choice == 'x_plus_y':
            integrand = x + coeff*y
        elif func_choice == 'xy':
            integrand = coeff * x * y
        else:
            integrand = coeff * x
        inner = integrate(integrand, (y, x**2, x))
        ans = integrate(inner, (x, 0, a))
        return problem(
            question=f"Evaluate $\\iint_R {latex(integrand)} \\, dA$ where $R$ is bounded by $y=x$ and $y=x^2$, $0 \\le x \\le {a}$.",
            answer=ans,
            difficulty=(1650, 1800),
            topic="multivariable_calculus/double_integrals",
            solution=steps(
                f"Region: $0 \\le x \\le {a}$, $x^2 \\le y \\le x$",
                f"$\\int_0^{{{a}}} \\int_{{x^2}}^x {latex(integrand)} \\, dy \\, dx$",
                f"Inner: $= {latex(inner)}$",
                f"Outer: $= {latex(ans)}$"
            ),
            calculator="scientific"
        )

    elif problem_type == 4:
        # Polar sector
        r_val = randint(2, 5)
        angle_frac = choice([Rational(1, 4), Rational(1, 3), Rational(1, 2), Rational(2, 3)])
        theta_max = angle_frac * pi
        integrand_r = r**2
        ans = integrate(integrate(r * integrand_r, (r, 0, r_val)), (t, 0, theta_max))
        if angle_frac == Rational(1, 4):
            angle_str = "\\frac{\\pi}{4}"
        elif angle_frac == Rational(1, 3):
            angle_str = "\\frac{\\pi}{3}"
        elif angle_frac == Rational(1, 2):
            angle_str = "\\frac{\\pi}{2}"
        else:
            angle_str = "\\frac{2\\pi}{3}"
        return problem(
            question=f"Evaluate $\\iint_R r^2 \\, dA$ over $0 \\le r \\le {r_val}$, $0 \\le \\theta \\le {angle_str}$.",
            answer=ans,
            difficulty=(1700, 1850),
            topic="multivariable_calculus/double_integrals",
            solution=steps(
                f"$dA = r \\, dr \\, d\\theta$",
                f"$\\int_0^{{{angle_str}}} \\int_0^{{{r_val}}} r^3 \\, dr \\, d\\theta = \\int_0^{{{angle_str}}} \\frac{{{r_val**4}}}{{4}} \\, d\\theta$",
                f"$= \\frac{{{r_val**4}}}{{4}} \\cdot {angle_str} = {latex(ans)}$"
            ),
            calculator="scientific"
        )

    else:
        # Type I region with variable bounds
        b = randint(2, 4)
        c = nonzero(-2, 2)
        func_choice = choice(['x_plus_cy', 'xy', 'cx2'])
        if func_choice == 'x_plus_cy':
            integrand = x + c*y
        elif func_choice == 'xy':
            integrand = c * x * y
        else:
            integrand = c * x**2
        a_lim = randint(1, 2)
        inner = integrate(integrand, (x, y, b))
        ans = integrate(inner, (y, 0, a_lim))
        return problem(
            question=f"Evaluate $\\int_{{0}}^{{{a_lim}}} \\int_{{y}}^{{{b}}} {latex(integrand)} \\, dx \\, dy$",
            answer=ans,
            difficulty=(1600, 1750),
            topic="multivariable_calculus/double_integrals",
            solution=steps(
                f"Region: $0 \\le y \\le {a_lim}$, $y \\le x \\le {b}$",
                f"Inner: $\\int_y^{{{b}}} {latex(integrand)} \\, dx = {latex(inner)}$",
                f"Outer: $= {latex(ans)}$"
            ),
            calculator="scientific"
        )

emit(generate())

File: src/multivariable_calculus_double_integrals_hard.jl
# multivariable_calculus - double_integrals (hard)
# Generated: 2026-03-08T21:01:44.819133

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y begin
    set_topic!("multivariable_calculus/double_integrals")
    
    problem_type = choice(1:5)
    
    if problem_type == 1
        # Type 1: Double integral with polynomial integrand over rectangular region
        # ELO: 1800-2000
        a = randint(1, 4)
        b = randint(1, 4)
        m = randint(1, 3)
        n = randint(1, 3)
        
        x_lo = randint(-3, 0)
        x_hi = randint(1, 4)
        y_lo = randint(-3, 0)
        y_hi = randint(1, 4)
        
        integrand = a*x^m + b*y^n
        
        # Compute answer by hand
        x_antideriv = a*x^(m+1)//(m+1)
        x_eval = substitute(x_antideriv, x => x_hi) - substitute(x_antideriv, x => x_lo)
        
        y_antideriv = b*y^(n+1)//(n+1)
        y_eval = substitute(y_antideriv, y => y_hi) - substitute(y_antideriv, y => y_lo)
        
        ans_numerator = (x_hi^(m+1) - x_lo^(m+1))*(y_hi^(n+1) - y_lo^(n+1))
        ans_denom = (m+1)*(n+1)
        
        if m == n
            ans = a*x_eval*(y_hi - y_lo) + b*(x_hi - x_lo)*y_eval
        else
            # Use symbolic computation
            temp1 = substitute(integrand, x => x)
            integrated_y = simplify(substitute(b*y^(n+1)//(n+1), y => y_hi) - substitute(b*y^(n+1)//(n+1), y => y_lo))
            integrated_x = simplify(substitute(a*x^(m+1)//(m+1), x => x_hi) - substitute(a*x^(m+1)//(m+1), x => x_lo))
            
            ans = a*integrated_x*(y_hi - y_lo) + b*(x_hi - x_lo)*integrated_y
        end
        
        problem(
            question="Evaluate \\\$\\displaystyle \\int_{$(x_lo)}^{$(x_hi)} \\int_{$(y_lo)}^{$(y_hi)} ($(tex(integrand))) \\, dy \\, dx\\\$",
            answer=ans,
            difficulty=(1800, 2000),
            solution=steps(
                sol("Given", "\\int_{$(x_lo)}^{$(x_hi)} \\int_{$(y_lo)}^{$(y_hi)} ($(tex(integrand))) \\, dy \\, dx"),
                "Integrate with respect to y first: \\\$\\int_{$(y_lo)}^{$(y_hi)} ($(tex(integrand))) \\, dy = $(a*x^m*(y_hi - y_lo)) + $(tex(y_eval))\\\$",
                "Now integrate with respect to x",
                sol("Answer", ans)
            ),
            time=180
        )
        
    elseif problem_type == 2
        # Type 2: Double integral with triangular region (Type I or II)
        # ELO: 2000-2200
        use_type_1 = choice([true, false])
        
        if use_type_1
            # Type I: integrate dy dx, bounds y from g1(x) to g2(x)
            x_lo = randint(0, 2)
            x_hi = randint(3, 6)
            m1 = randint(1, 3)
            b1 = randint(-2, 1)
            
            integrand_coeff = nonzero(-5, 5)
            
            # Region: y from 0 to m1*x + b1, x from x_lo to x_hi
            y_upper = m1*x + b1
            
            integrand = integrand_coeff*x*y
            
            # Integrate y first
            y_int = substitute(integrand_coeff*x*y^2//2, y => y_upper) - 0
            y_int = expand(y_int)
            
            # Then integrate x
            final_int = simplify(substitute(diff(y_int, x) * 0, x => x) + y_int)
            
            # Manual calculation for clean answer
            # ∫∫ c*x*y dy dx = ∫ c*x*(m1*x+b1)^2/2 dx from x_lo to x_hi
            # = c/2 ∫ x*(m1^2*x^2 + 2*m1*b1*x + b1^2) dx
            # = c/2 ∫ (m1^2*x^3 + 2*m1*b1*x^2 + b1^2*x) dx
            # = c/2 [m1^2*x^4/4 + 2*m1*b1*x^3/3 + b1^2*x^2/2] from x_lo to x_hi
            
            term1 = m1^2*(x_hi^4 - x_lo^4)//4
            term2 = 2*m1*b1*(x_hi^3 - x_lo^3)//3
            term3 = b1^2*(x_hi^2 - x_lo^2)//2
            
            ans = integrand_coeff*(term1 + term2 + term3)//2
            
            problem(
                question="Evaluate \\\$\\displaystyle \\int_{$(x_lo)}^{$(x_hi)} \\int_{0}^{$(tex(y_upper))} $(tex(integrand)) \\, dy \\, dx\\\$",
                answer=ans,
                difficulty=(2000, 2200),
                solution=steps(
                    sol("Given", "\\int_{$(x_lo)}^{$(x_hi)} \\int_{0}^{$(tex(y_upper))} $(tex(integrand)) \\, dy \\, dx"),
                    "Integrate with respect to y: \\\$\\int_0^{$(tex(y_upper))} $(tex(integrand)) \\, dy = $(tex(integrand_coeff*x*(y_upper)^2//2))\\\$",
                    "Expand and integrate with respect to x",
                    sol("Answer", ans)
                ),
                time=240
            )
        else
            # Type II: integrate dx dy
            y_lo = randint(0, 2)
            y_hi = randint(3, 6)
            m2 = randint(1, 3)
            
            integrand_coeff = nonzero(-5, 5)
            
            # Region: x from 0 to y/m2, y from y_lo to y_hi
            x_upper = y//m2
            
            integrand = integrand_coeff*x*y
            
            # Manual calculation
            # ∫∫ c*x*y dx dy = ∫ c*y*(y/m2)^2/2 dy from y_lo to y_hi
            # = c/(2*m2^2) ∫ y^3 dy
            # = c/(2*m2^2) * [y^4/4] from y_lo to y_hi
            
            ans = integrand_coeff*(y_hi^4 - y_lo^4)//(8*m2^2)
            
            problem(
                question="Evaluate \\\$\\displaystyle \\int_{$(y_lo)}^{$(y_hi)} \\int_{0}^{y/$(m2)} $(tex(integrand)) \\, dx \\, dy\\\$",
                answer=ans,
                difficulty=(2000, 2200),
                solution=steps(
                    sol("Given", "\\int_{$(y_lo)}^{$(y_hi)} \\int_{0}^{y/$(m2)} $(tex(integrand)) \\, dx \\, dy"),
                    "Integrate with respect to x: \\\$\\int_0^{y/$(m2)} $(tex(integrand)) \\, dx = $(tex(integrand_coeff*y*(y//m2)^2//2))\\\$",
                    "Simplify and integrate with respect to y",
                    sol("Answer", ans)
                ),
                time=240
            )
        end
        
    elseif problem_type == 3
        # Type 3: Change order of integration
        # ELO: 2100-2300
        a = randint(1, 5)
        b = randint(1, 4)
        
        # Original: ∫[0,a] ∫[x,a] f(x,y) dy dx
        # New: ∫[0,a] ∫[0,y] f(x,y) dx dy
        
        x_lo = 0
        x_hi = a
        
        integrand = nonzero(-3, 3)*x + nonzero(-3, 3)*y
        
        # For grading, we'll accept the new integral bounds
        new_y_lo = 0
        new_y_hi = a
        new_x_lo_expr = 0
        new_x_hi_expr = y
        
        ans_text = "\\int_{0}^{$(a)} \\int_{0}^{y} ($(tex(integrand))) \\, dx \\, dy"
        
        problem(
            question="Reverse the order of integration for \\\$\\displaystyle \\int_{0}^{$(a)} \\int_{x}^{$(a)} ($(tex(integrand))) \\, dy \\, dx\\\$",
            answer=ans_text,
            difficulty=(2100, 2300),
            answer_type="expression",
            solution=steps(
                "Sketch the region: \\\$0 \\le x \\le $(a)\\\$, \\\$x \\le y \\le $(a)\\\$",
                "This is the region above the line \\\$y = x\\\$ and below \\\$y = $(a)\\\$",
                "Reversing order: \\\$0 \\le y \\le $(a)\\\$, \\\$0 \\le x \\le y\\\$",
                sol("Answer", ans_text)
            ),
            time=180
        )
        
    elseif problem_type == 4
        # Type 4: Polar coordinates double integral
        # ELO: 2200-2400
        r_hi = randint(2, 5)
        
        # Full circle or sector
        use_full = choice([true, false])
        
        if use_full
            theta_lo = 0
            theta_hi = 2
            integrand = 1  # Just area
            
            ans = "\\pi \\cdot $(r_hi)^2"
            ans_numeric = r_hi^2
            
            problem(
                question="Evaluate \\\$\\displaystyle \\int_{0}^{2\\pi} \\int_{0}^{$(r_hi)} r \\, dr \\, d\\theta\\\$ (Express answer as a multiple of π)",
                answer=ans_numeric,
                difficulty=(2000, 2200),
                solution=steps(
                    sol("Given", "\\int_{0}^{2\\pi} \\int_{0}^{$(r_hi)} r \\, dr \\, d\\theta"),
                    "Integrate with respect to r: \\\$\\int_0^{$(r_hi)} r \\, dr = \\frac{r^2}{2}\\Big|_0^{$(r_hi)} = \\frac{$(r_hi^2)}{2}\\\$",
                    "Integrate with respect to θ: \\\$\\int_0^{2\\pi} \\frac{$(r_hi^2)}{2} \\, d\\theta = $(r_hi^2)\\pi\\\$",
                    sol("Answer", "$(r_hi^2)\\pi")
                ),
                time=150
            )
        else
            # Sector
            theta_hi = choice([1, 2, 3, 4])  # Multiples of π
            integrand_r = choice([1, 2])
            
            # ∫∫ r^integrand_r dr dθ
            # = ∫ r^(integrand_r+1)/(integrand_r+1) |_0^r_hi dθ
            # = ∫ r_hi^(integrand_r+1)/(integrand_r+1) dθ from 0 to theta_hi*π
            
            r_integral = r_hi^(integrand_r+1)//(integrand_r+1)
            ans_numeric = r_integral * theta_hi
            
            if integrand_r == 1
                integrand_tex = "r"
            else
                integrand_tex = "r^$(integrand_r)"
            end
            
            problem(
                question="Evaluate \\\$\\displaystyle \\int_{0}^{$(theta_hi)\\pi} \\int_{0}^{$(r_hi)} $(integrand_tex) \\, dr \\, d\\theta\\\$ (Express answer as a multiple of π)",
                answer=ans_numeric,
                difficulty=(2100, 2300),
                solution=steps(
                    sol("Given", "\\int_{0}^{$(theta_hi)\\pi} \\int_{0}^{$(r_hi)} $(integrand_tex) \\, dr \\, d\\theta"),
                    "Integrate with respect to r: \\\$\\int_0^{$(r_hi)} $(integrand_tex) \\, dr = \\frac{r^{$(integrand_r+1)}}{$(integrand_r+1)}\\Big|_0^{$(r_hi)} = $(tex(r_integral))\\\$",
                    "Integrate with respect to θ: \\\$\\int_0^{$(theta_hi)\\pi} $(tex(r_integral)) \\, d\\theta = $(tex(ans_numeric))\\pi\\\$",
                    sol("Answer", "$(tex(ans_numeric))\\pi")
                ),
                time=180
            )
        end
        
    else
        # Type 5: Volume under surface (challenging setup)
        # ELO: 2300-2500
        a = randint(2, 4)
        b = randint(2, 4)
        c = nonzero(1, 3)
        
        # Volume under z = c - x - y over rectangle [0,a] × [0,b]
        # V = ∫∫ (c - x - y) dA where c > a + b to keep positive
        c = a + b + randint(1, 3)
        
        # ∫[0,a] ∫[0,b] (c - x - y) dy dx
        # = ∫[0,a] [c*y - x*y - y^2/2]|_0^b dx
        # = ∫[0,a] (c*b - x*b - b^2/2) dx
        # = [c*b*x - b*x^2/2 - b^2*x/2]|_0^a
        # = c*b*a - b*a^2/2 - b^2*a/2
        
        ans = c*b*a - b*a^2//2 - b^2*a//2
        
        problem(
            question="Find the volume of the solid bounded by z = $(c) - x - y and the coordinate planes over [0, $(a)] x [0, $(b)].",
            answer=ans,
            difficulty=(2300, 2500),
            solution=steps(
                "Set up double integral",
                "Inner integral with respect to y, then outer with respect to x",
                sol("Answer", ans)
            ),
            time=210
        )
    end
end
